### PR TITLE
Adds a logout route to state that executes the logout code

### LINKED ISSUE #2617: GUI needs to support /logout route as a root state
We need the /logout route to be visit-able directly on the gui--it should be part of state and execute the gui logout code and issue any necessary redirects, e.g. IdM or gisf logout.

### PR DESCRIPTION
**Pull request by jaycee**
Fixes #2617

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7779/

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -908,17 +908,16 @@ YUI.add('juju-gui', function(Y) {
...
-        logout={this.logout.bind(this)}
-        clearCookie={bakery.storage.clear.bind(bakery.storage)}
-        gisfLogout={window.juju_config.gisfLogout || ''}
-        gisf={window.juju_config.gisf || false}
         charmstoreLogoutUrl={charmstore.getLogoutUrl()}
-        getUser={this.getUser.bind(this, 'charmstore')}
-        clearUser={this.clearUser.bind(this, 'charmstore')}
+        doCharmstoreLogout={doCharmstoreLogout}
+        locationAssign={window.location.assign}
+        logoutUrl={'/logout'}
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: If we're hardcoding it here, why don't we hard code it in the component and save the prop?

**Comment by jaycee** [comment]
QA:

Using the logout link should work normally, both in gisf and in gijoe. In gijoe be sure to test that logging out works once you've logged into the charmstore as well.

Also test that going to /logout directly works. NB: it won't trigger proper charmstore logout, since that requires hitting the api logout url directly.

File: jujugui/static/gui/src/app/app.js [hunk 2]
```diff
@@ -2113,7 +2112,6 @@ YUI.add('juju-gui', function(Y) {
    */
     _ensureControllerConnection: function(state, next) {
       if (
-        state.root !== 'logout' &&
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: hmm, why isn't this necessary? I know we kind of wrote this part assuming it was, just curious.

File: jujugui/static/gui/src/app/app.js [hunk 3]
```diff
@@ -2600,21 +2603,12 @@ YUI.add('juju-gui', function(Y) {
       }
     },
 
-    // Route handlers
+    _handleLogout: function() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: docbloc plz

File: jujugui/static/gui/src/app/app.js [hunk 4]
```diff
@@ -2199,6 +2197,11 @@ YUI.add('juju-gui', function(Y) {
           this._renderLogin();
           next();
           break;
+        case 'logout':
+          this._handleLogout();
+          this.state.changeState({root: 'login'});
```
Comment (hatched) [inline jujugui/static/gui/src/app/app.js]: If we're in gisf the code below navigates away from the page. Could you add a comment here to that effect as it'll be easily missed in the future.

File: jujugui/static/gui/src/app/components/logout/logout.js [hunk 5]
```diff
@@ -70,16 +45,27 @@ YUI.add('logout-component', function() {
...
+    },
+
     render: function() {
-      let href = this.props.charmstoreLogoutUrl ;
-      if (this.props.gisf) {
-        href = this.props.gisfLogout;
+      const props = this.props;
+      let logoutUrl = props.logoutUrl;
+      let target = '_self';
+      if (props.doCharmstoreLogout()) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/logout/logout.js]: the props shows this as required so this check isn't really necessary.

File: jujugui/static/gui/src/app/app.js (hunk 2 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: It's actually actively bad--the logout mechanisms need an active websocket. So we need to have a connection to logout. Which is peculiar, but not actually bad.

File: jujugui/static/gui/src/app/components/logout/logout.js (hunk 5 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/components/logout/logout.js]: That's a function call; returns true or false.

Comment (hatched) [inline jujugui/static/gui/src/app/components/logout/logout.js]: doh....sorry :)

**Comment by makyo** [comment]
Looks good, QAing now.

**Comment by makyo** [comment]
Running into some problems on gijoe:

When I visit https://xx.xx.xx.xx:17070/gui/u/admin/default/ and click logout, it takes me to https://xx.xx.xx.xx:17070/logout, which isn't a path gijoe recognizes.

Additionally, when I click logout when logged into the charm store, it takes me to the logout page, but doesn't seem to clear the creds in the app, so the link never changes to /logout, just stays the charmstore logout url. Going over the steps again to make sure I'm doing this right, though.

File: jujugui/static/gui/src/app/app.js (hunk 1 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Fair enough. I've changed it.

File: jujugui/static/gui/src/app/app.js (hunk 4 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Done.

File: jujugui/static/gui/src/app/app.js (hunk 3 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Done.

File: jujugui/static/gui/src/app/components/logout/logout.js (hunk 5 above)
Comment (jaycee) [inline jujugui/static/gui/src/app/components/logout/logout.js]: No worries. :0

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7780/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7781/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/7782/

**Comment by makyo** [comment]
+1

**Comment by jaycee** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**